extend_foreach_to_cover_generate_rules.md

```diff
@@ -0,0 +1,182 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+
+# Overview
+[overview]: #overview
+Generate rules are in charge of creating new resources. The rule can either define the new resource using the `data` object or use an existing resource in the cluster with the `clone` object. Currently, only one resource can be generated. 
+
+This proposal is to extend the `forEach` to cover generate rules.
+
+# Definitions
+[definitions]: #definitions
+
+1. Trigger: It refers to the resource responsible for triggering the generate rule as defined in a combination of `match` and `exclude` blocks.
+
+2. Downstream: It refers to the generated resource(s).
+
+3. Source: It refers to the clone source when generating new resources from existing ones in the cluster.
+
+# Motivation
+[motivation]: #motivation
+
+There are some use cases where a single "trigger" resource should result in the creation of multiple downstream resources.
+
+Examples:
+1. Generating a list of new network policies based on a comma separated string in an annotation using a `forEach` and `Split()`.
+2. Migrating from ingress to Kubernetes Gateway API requires generating at least two resources: `Gateway` and `HTTPRoute`.
+
+# Proposal
+In this proposal, Kyverno generate rule can be extended to allowing the generation of multiple resources using `forEach`.
+This could be done by using the `list` attribute which is a JMESPath expression that defines the sub-elements it processes.
+
+Examples:
+1. Generate multiple network policies when a new namespace is created:
+```yaml
+spec:
+  rules:
+  - name: generate-network-policies
+    match:
+      any:
+      - resources:
+          kinds:
+          - Namespace
+    generate:
+      synchronize: true
+      foreach:
+        - list: request.object.metadata.labels.networkpolicies | split(@, ',')
+          resources:
+            - apiVersion: v1
+              kind: NetworkPolicy
+              name: '{{ element }}'
+              namespace: "{{request.object.metadata.name}}"
+              data:
+                spec:
+                  podSelector: {}
+                  policyTypes:
+                  - Ingress
+                  - Egress
+```
+
+2. Generate ConfigMap and Secret for each container:
+```yaml
+spec:
+  rules:
+  - name: generate-configmaps-and-secrets
+    match:
+      any:
+      - resources:
+          kinds:
+          - Pod
+    generate:
+      synchronize: true
+      foreach:
+        - list: request.object.spec.containers
+          resources:
+            - apiVersion: v1
+              kind: ConfigMap
+              name: cm
+              namespace: "{{request.object.metadata.namespace}}"
+              data:
+                foo: bar
+            - apiVersion: v1
+              kind: Secret
+              name: secret
+              namespace: "{{request.object.metadata.namespace}}"
+              data:
```

Comment:
Do we support `clone` and `cloneList` in `foreach`?

Reply:
No. Since `cloneList` is used to clone multiple resources at once, then there is no need for using `foreach`. WDYT?

Reply:
What about `clone`?

Reply:
Using `foreach` with `clone` will allow us to clone the existing resource multiple times.
Do we have a real use case for that?

For example, we can clone a secret depending on the number of containers. For each container, the same secret will be cloned.

Reply:
Would this handle the case of cloning a resource into multiple namespaces? 

For example matching on `Secret` with a label of the form `sync: a=b`, using the API server to find all `Namespace` with label `a: b` and then `foreach`'ing that list to generate the Secrets.

This is a case we can't currently do with Kyverno (so far as I could tell). It's a task tools like KubeD or Reflector perform as their primary purpose. (Along with many others) I'm looking for a replacement for the former and was hoping I'd be able to write a `ClusterPolicy` for that case... but it seems I'd need `foreach` in a `generate` rule that could also `clone` _first_

Reply:
You might consider using external-secrets kubernetes provider along with its PushSecret custom resource to replace many kubed use cases.

Reply:
Hi @pinkfloydx33 - this is an interesting use case. 

I have crafted an example rule which looks up a list of namespaces by the label and clone the secret to multiple namespaces.
```yaml
spec:
  rules:
  - name: clone-seret-to-multiple-namespace
    match:
      any:
      - resources:
          kinds:
          - Secret
          namespaces:
          - default
    context:
    - variable:
        name: selector
        jmesPath: request.object.metadata.labels.sync
        default: a=b
    - name: namespaces
      apiCall:
        urlPath: /api/v1/namespaces?labelSelector={{selector}}?limit=5
    generate:
      synchronize: false
      foreach:
        - list: "{{namespaces}}"
          cloneList:
          - source:
              namespace: default
              name: regcred
              kind: v1
              apiVersion: Secret
            target:
              namespace: "{{element}}"
              name: regcred
```

What do you think?